File: packages/sdk-go/validation.go
package taskforceai

import (
	"fmt"
	"strings"
)

func validateTaskID(taskID, label string) error {
	if strings.TrimSpace(taskID) == "" {
		return fmt.Errorf("invalid %s response: taskId is required", label)
	}
	return nil
}

func validateTaskStatus(status TaskStatus, label string) error {
	if err := validateTaskID(status.TaskID, label); err != nil {
		return err
	}
	switch status.Status {
	case "processing", "completed", "failed", "awaiting_approval":
		return nil
	default:
		return fmt.Errorf("invalid %s response: unsupported status %q", label, status.Status)
	}
}

func validateThread(thread Thread, label string) error {
	if thread.ID <= 0 {
		return fmt.Errorf("invalid %s response: id must be positive", label)
	}
	if strings.TrimSpace(thread.Title) == "" {
		return fmt.Errorf("invalid %s response: title is required", label)
	}
	if thread.CreatedAt.IsZero() {
		return fmt.Errorf("invalid %s response: created_at is required", label)
	}
	if thread.UpdatedAt.IsZero() {
		return fmt.Errorf("invalid %s response: updated_at is required", label)
	}
	return nil
}

func validateThreadList(result ThreadListResponse, label string) error {
	if result.Total < 0 {
		return fmt.Errorf("invalid %s response: total must be non-negative", label)
	}
	for index, thread := range result.Threads {
		if err := validateThread(thread, fmt.Sprintf("%s thread %d", label, index)); err != nil {
			return err
		}
	}
	return nil
}

func validateThreadMessage(message ThreadMessage, label string) error {
	if message.ID <= 0 {
		return fmt.Errorf("invalid %s response: id must be positive", label)
	}
	if message.ThreadID <= 0 {
		return fmt.Errorf("invalid %s response: thread_id must be positive", label)
	}
	if message.Role != "user" && message.Role != "assistant" {
		return fmt.Errorf("invalid %s response: unsupported role %q", label, message.Role)
	}
	if message.CreatedAt.IsZero() {
		return fmt.Errorf("invalid %s response: created_at is required", label)
	}
	return nil
}

func validateThreadMessages(result ThreadMessagesResponse, label string) error {
	if result.Total < 0 {
		return fmt.Errorf("invalid %s response: total must be non-negative", label)
	}
	for index, message := range result.Messages {
		if err := validateThreadMessage(message, fmt.Sprintf("%s message %d", label, index)); err != nil {
			return err
		}
	}
	return nil
}

func validateThreadRun(result ThreadRunResponse, label string) error {
	if strings.TrimSpace(result.TaskID) == "" {
		return fmt.Errorf("invalid %s response: task_id is required", label)
	}
	if result.ThreadID <= 0 {
		return fmt.Errorf("invalid %s response: thread_id must be positive", label)
	}
	if result.MessageID <= 0 {
		return fmt.Errorf("invalid %s response: message_id must be positive", label)
	}
	return nil
}

func validateFile(file File, label string) error {
	if strings.TrimSpace(file.ID) == "" {
		return fmt.Errorf("invalid %s response: id is required", label)
	}
	if strings.TrimSpace(file.Filename) == "" {
		return fmt.Errorf("invalid %s response: filename is required", label)
	}
	if strings.TrimSpace(file.Purpose) == "" {
		return fmt.Errorf("invalid %s response: purpose is required", label)
	}
	if file.Bytes < 0 {
		return fmt.Errorf("invalid %s response: bytes must be non-negative", label)
	}
	if file.CreatedAt.IsZero() {
		return fmt.Errorf("invalid %s response: created_at is required", label)
	}
	return nil
}

func validateFileList(result FileListResponse, label string) error {
	if result.Total < 0 {
		return fmt.Errorf("invalid %s response: total must be non-negative", label)
	}
	for index, file := range result.Files {
		if err := validateFile(file, fmt.Sprintf("%s file %d", label, index)); err != nil {
			return err
		}
	}
	return nil
}

File: packages/sdk-rust/src/validation.rs
use crate::error::TaskForceAIError;
use crate::files::{File, FileListResponse};
use crate::threads::{Thread, ThreadListResponse, ThreadMessagesResponse, ThreadRunResponse};
use crate::types::{SubmitTaskResponse, TaskStatus};

fn required_string(value: &str, label: &str, field: &str) -> Result<(), TaskForceAIError> {
    if value.trim().is_empty() {
        return Err(TaskForceAIError::Validation(format!(
            "invalid {label} response: {field} is required"
        )));
    }
    Ok(())
}

fn positive_i64(value: i64, label: &str, field: &str) -> Result<(), TaskForceAIError> {
    if value <= 0 {
        return Err(TaskForceAIError::Validation(format!(
            "invalid {label} response: {field} must be positive"
        )));
    }
    Ok(())
}

fn non_negative_i64(value: i64, label: &str, field: &str) -> Result<(), TaskForceAIError> {
    if value < 0 {
        return Err(TaskForceAIError::Validation(format!(
            "invalid {label} response: {field} must be non-negative"
        )));
    }
    Ok(())
}

pub(crate) fn validate_submit_task_response(
    response: &SubmitTaskResponse,
) -> Result<(), TaskForceAIError> {
    required_string(&response.task_id, "task submission", "taskId")
}

pub(crate) fn validate_task_status(status: &TaskStatus) -> Result<(), TaskForceAIError> {
    required_string(&status.task_id, "task status", "taskId")
}

pub(crate) fn validate_file(file: &File, label: &str) -> Result<(), TaskForceAIError> {
    required_string(&file.id, label, "id")?;
    required_string(&file.filename, label, "filename")?;
    required_string(&file.purpose, label, "purpose")?;
    non_negative_i64(file.bytes, label, "bytes")
}

pub(crate) fn validate_file_list(response: &FileListResponse) -> Result<(), TaskForceAIError> {
    non_negative_i64(response.total, "file list", "total")?;
    for (index, file) in response.files.iter().enumerate() {
        validate_file(file, &format!("file list item {index}"))?;
    }
    Ok(())
}

pub(crate) fn validate_thread(thread: &Thread, label: &str) -> Result<(), TaskForceAIError> {
    positive_i64(thread.id, label, "id")?;
    required_string(&thread.timestamp, label, "timestamp")?;
    non_negative_i64(thread.execution_time, label, "execution_time")?;
    non_negative_i64(thread.agent_count, label, "agent_count")
}

pub(crate) fn validate_thread_list(response: &ThreadListResponse) -> Result<(), TaskForceAIError> {
    non_negative_i64(response.total, "thread list", "total")?;
    non_negative_i64(response.limit, "thread list", "limit")?;
    non_negative_i64(response.offset, "thread list", "offset")?;
    for (index, thread) in response.conversations.iter().enumerate() {
        validate_thread(thread, &format!("thread list item {index}"))?;
    }
    Ok(())
}

pub(crate) fn validate_thread_messages(
    response: &ThreadMessagesResponse,
) -> Result<(), TaskForceAIError> {
    for (index, message) in response.messages.iter().enumerate() {
        let label = format!("thread message {index}");
        positive_i64(message.id, &label, "id")?;
        positive_i64(message.thread_id, &label, "thread_id")?;
        if message.role != "user" && message.role != "assistant" {
            return Err(TaskForceAIError::Validation(format!(
                "invalid {label} response: role is unsupported"
            )));
        }
    }
    Ok(())
}

pub(crate) fn validate_thread_run(response: &ThreadRunResponse) -> Result<(), TaskForceAIError> {
    required_string(&response.task_id, "thread run", "taskId")?;
    required_string(&response.status, "thread run", "status")
}
